src/XtensaOptimize.cpp

```diff
@@ -2063,6 +2063,284 @@ class SimplifySliceConcat : public IRGraphMutator {
+        // Store current loop in the scope.
+        loops.emplace_back(op->for_type == ForType::Parallel);
+        loops.back().name = op->name;
+        loops.back().min = op->min;
+        loops.back().extent = op->extent;
+        loops.back().loop_index = loop_index++;
+
+        // On the first pass we will compute a list of allocations which are not
+        // allowed to be lifted to this loop level/
+        if (first_run) {
+            blocked_allocations.emplace_back();
+        }
+
+        Stmt for_op = IRMutator::visit(op);
+        // Move allocations by wrapping them around this loop.
+        if (first_run) {
+            internal_assert(loops.back().allocations_to_move.empty()) << "allocation_to_move list is expected to be empty after the first pass";
+        }
+        if (!loops.back().allocations_to_move.empty()) {
+            debug(3) << "Starting to move allocations "
+                     << loops.back().name << " "
+                     << loops.back().allocations_to_move.size() << "\n";
+            for (const auto &a : loops.back().allocations_to_move) {
+                const auto &alloc = a.second;
+                debug(3) << "Moving allocation - " << alloc.name << "\n";
+                Expr total_extent = alloc.total_extent;
+                if (alloc.const_allocation_size > 0) {
+                    total_extent = alloc.const_allocation_size;
+                }
+                for_op = Allocate::make(alloc.name, alloc.type, alloc.memory_type,
+                                        {total_extent}, alloc.condition, for_op,
+                                        alloc.new_expr, alloc.free_function,
+                                        alloc.padding);
+            }
+            debug(3) << "Finished moving allocations : \n"
+                     << ""
+                     << "\n";
+        }
+        loops.pop_back();
+        return for_op;
+    }
+
+    Stmt visit(const Block *block) override {
+        if (const Free *maybe_free = block->first.as<Free>()) {
+            for (const auto &loop : loops) {
+                if (loop.allocations_to_move.count(maybe_free->name) > 0) {
+                    return mutate(block->rest);
+                }
+            }
+        }
+        if (const Free *maybe_free = block->rest.as<Free>()) {
+            for (const auto &loop : loops) {
+                if (loop.allocations_to_move.count(maybe_free->name) > 0) {
+                    return mutate(block->first);
+                }
+            }
+        }
+
+        return IRMutator::visit(block);
+    }
+
+    Stmt visit(const Allocate *op) override {
+        if (!loops.empty() &&
+            (op->memory_type == MemoryType::Stack || op->memory_type == MemoryType::VTCM) && !op->new_expr.defined() &&
+            (op->constant_allocation_size() == 0 ||
+             // Don't lift small allocations, constant-sized allocations.
+             op->constant_allocation_size() > 512)) {
```

Comment:
512 seems somewhat arbitrary. Why not lift all constant-sized allocations? I'm not clear on what the downside is.